### PR TITLE
Feat/flowbit prefilter/v26

### PR DESCRIPTION
**Pull request by victorjulien**
https://redmine.openinfosecfoundation.org/issues/2486

Rebase of #12380, also addresses some of the issues.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2317

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 24875

File: src/detect-engine-prefilter.c [hunk 1]
```diff
@@ -342,6 +371,39 @@ int PrefilterAppendFrameEngine(DetectEngineCtx *de_ctx, SigGroupHead *sgh,
...
+        return -1;
+    memset(e, 0x00, sizeof(*e));
+    e->PrefilterPostRule = PrefilterPostRuleFunc;
+    e->pectx = pectx;
+    e->Free = FreeFunc;
+
+    if (sgh->init->post_rule_match_engines == NULL) {
+        sgh->init->post_rule_match_engines = e;
+    } else {
+        PrefilterEngineList *t = sgh->init->post_rule_match_engines;
```
Comment (jlucovsky) [inline src/detect-engine-prefilter.c:393]: Is this list expected to be more than a few items? If so, would a list/queue that allows tail adds be more efficient?

File: src/detect-engine-prefilter.h [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2016 Open Information Security Foundation
+/* Copyright (C) 2016-2020 Open Information Security Foundation
```
Comment (jlucovsky) [inline src/detect-engine-prefilter.h:1]: nit: s/2020/2025

File: src/detect-flowbits.c [hunk 3]
```diff
@@ -132,11 +139,10 @@ static int FlowbitOrAddData(DetectEngineCtx *de_ctx, DetectFlowbitsData *cd, cha
...
 {
     if (p->flow == NULL)
-        return 0;
-
-    FlowBitToggle(p->flow,fd->idx);
+        return -1;
 
-    return 1;
+    bool set = FlowBitToggle(p->flow, fd->idx);
+    return set == true;
```
Comment (jlucovsky) [inline src/detect-flowbits.c:145]: Suggest `return FlowBitToggle()`

File: src/detect-flowbits.c [hunk 4]
```diff
@@ -809,6 +902,469 @@ static void DetectFlowbitsAnalyzeDump(const DetectEngineCtx *de_ctx,
...
+    for (const SigMatch *sm = s->init_data->smlists[DETECT_SM_LIST_MATCH]; sm != NULL;
+            sm = sm->next) {
+        switch (sm->type) {
+            case DETECT_FLOWBITS: {
+                const DetectFlowbitsData *fb = (DetectFlowbitsData *)sm->ctx;
+                if (fb->cmd == DETECT_FLOWBITS_CMD_ISSET) {
+                    SCLogDebug("sid:%u: FLOWBITS ISSET can prefilter", s->id);
+                    return true;
+                }
+            } break;
```
Comment (jlucovsky) [inline src/detect-flowbits.c:918]: nit: style

File: src/detect.c [hunk 5]
```diff
@@ -746,6 +746,16 @@ static bool IsOnlyTxInDirection(Flow *f, uint64_t txid, uint8_t dir)
     return false;
 }
 
+static int SortHelper(const void *a, const void *b)
+{
+    const Signature *sa = *(const Signature **)a;
+    const Signature *sb = *(const Signature **)b;
+    if (sa->num == sb->num)
+        return 0;
+    else
```
Comment (jlucovsky) [inline src/detect.c:755]: nit: else is unneeded.

File: src/flow-bit.c [hunk 6]
```diff
@@ -55,19 +55,25 @@ static FlowBit *FlowBitGet(Flow *f, uint32_t idx)
     return NULL;
 }
 
-/* add a flowbit to the flow */
-static void FlowBitAdd(Flow *f, uint32_t idx)
+/** \brief add a flowbit to the flow
+ *  \retval -1 error
+ *  \retval 0 not added, already set before
+ *  \retval 1 added */
```
Comment (catenacyber) [inline src/flow-bit.c:61]: wonder if using an enum would be better...

File: src/detect-engine-build.c [hunk 7]
```diff
@@ -668,7 +668,7 @@ static JsonBuilder *RulesGroupPrintSghStats(const DetectEngineCtx *de_ctx, const
             any5_cnt++;
         }
 
-        prefilter_cnt += (s->init_data->prefilter_sm != 0);
+        prefilter_cnt += (s->init_data->prefilter_sm != NULL);
```
Comment (catenacyber) [inline src/detect-engine-build.c:671]: nit own commit

**Comment by catenacyber** [comment]
> the prefilter is implemented as an RB_TREE of flowbits, with the rule id's they "enable" stored per tree node.

Why ?
Do flowbits have a unique id ?

**Comment by catenacyber** [comment]
> The 'isset' prefilter has one important corner case, which is that bits can in fact be set during the rule evaluation stage. This is different from all other prefilter engines, that evaluate an immutable state (for the lifetime of the packets inspection).

This looks strange and counterintuitive.
Do you mean that alerts like this will trigger on the same packet ?
```
alert http any any -> any any (http.uri; content:"test"; flowbits:set,uritest; sid:11;)
alert http any any -> any any (http.method; content:"HEAD"; flowbits:isset,uritest; sid:12;)
```

How does the engine know to run first the rules that set flowbits before the ones that need them ?

**Comment by victorjulien** [comment]
> > The 'isset' prefilter has one important corner case, which is that bits can in fact be set during the rule evaluation stage. This is different from all other prefilter engines, that evaluate an immutable state (for the lifetime of the packets inspection).
> 
> This looks strange and counterintuitive. Do you mean that alerts like this will trigger on the same packet ?
> 
> ```
> alert http any any -> any any (http.uri; content:"test"; flowbits:set,uritest; sid:11;)
> alert http any any -> any any (http.method; content:"HEAD"; flowbits:isset,uritest; sid:12;)
> ```
> 
> How does the engine know to run first the rules that set flowbits before the ones that need them ?

During rule ordering a rule that sets a bit will be put before a sigs that checks the bit.

**Comment by victorjulien** [comment]
> > the prefilter is implemented as an RB_TREE of flowbits, with the rule id's they "enable" stored per tree node.
> 
> Why ? Do flowbits have a unique id ? Could it not be a simple array ? (even if it is painful to build)

Most flowbits will not use this, so the id space is quite sparse. An array is going to be large.

**Comment by catenacyber** [comment]
> During rule ordering a rule that sets a bit will be put before a sigs that checks the bit.

How do you handle cycles ?

Like
```
alert ip any any -> any any (someconditions; flowbits:isset,bitC; flowbits:set,bitA; sid:11;)
alert ip any any -> any any (someconditions; flowbits:isset,bitA; flowbits:set,bitB; sid:12;)
alert ip any any -> any any (someconditions; flowbits:isset,bitB; flowbits:set,bitC; sid:13;)

alert ip any any -> any any (someconditions; flowbits:set,bitA; sid:9;)
alert ip any any -> any any (someconditions; flowbits:set,bitB; sid:8;)
alert ip any any -> any any (someconditions; flowbits:set,bitC; sid:7;)
```

**Comment by catenacyber** [comment]
> Most flowbits will not use this, so the id space is quite sparse. An array is going to be large.

Indeed, but are not a few kilobytes of RAM worth the nanoseconds for direct access ?

I think prefilter is about optimizing speed, right ?
For the structure, we want to optimize getting an element, right ?
There are not so many flowbits, so I think we can afford a big sparse array as it brings speed improvement, whereas a rb tree takes O(log(n)) steps to get an element

**Comment by victorjulien** [comment]
> > During rule ordering a rule that sets a bit will be put before a sigs that checks the bit.
> 
> How do you handle cycles ?
> 
> Like
> 
> ```
> alert ip any any -> any any (someconditions; flowbits:isset,bitC; flowbits:set,bitA; sid:11;)
> alert ip any any -> any any (someconditions; flowbits:isset,bitA; flowbits:set,bitB; sid:12;)
> alert ip any any -> any any (someconditions; flowbits:isset,bitB; flowbits:set,bitC; sid:13;)
> 
> alert ip any any -> any any (someconditions; flowbits:set,bitA; sid:9;)
> alert ip any any -> any any (someconditions; flowbits:set,bitB; sid:8;)
> alert ip any any -> any any (someconditions; flowbits:set,bitC; sid:7;)
> ```

This isn't new, the ordering code exists for a long time already.

**Comment by victorjulien** [comment]
> > Most flowbits will not use this, so the id space is quite sparse. An array is going to be large.
> 
> Indeed, but are not a few kilobytes of RAM worth the nanoseconds for direct access ?
> 
> I think prefilter is about optimizing speed, right ? For the structure, we want to optimize getting an element, right ? There are not so many flowbits, so I think we can afford a big sparse array as it brings speed improvement, whereas a rb tree takes O(log(n)) steps to get an element

This discussion is getting ahead of itself. If you want to argue about the performance you'll have to bring profiling results, not just theory to the table here.

**Comment by catenacyber** [comment]
> This isn't new, the ordering code exists for a long time already.

Sure.
For reviewing this PR, I just want to understand this flowbits behavior as this PR commit message highlights the importance of this edge case (that I did not know of)

**Comment by catenacyber** [comment]
> This discussion is getting ahead of itself. If you want to argue about the performance you'll have to bring profiling results, not just theory to the table here.

I am just discussing the design, I did not even look at the code.

I can do a benchmark to show that `PFB_RB_FIND(&ctx->fb_tree, &lookup);` is slower than `ctx->fb_array[gv->idx];` but is this what we want to achieve ? (optimize speed of `PrefilterFlowbitMatch` )

**Comment by victorjulien** [comment]
> > This discussion is getting ahead of itself. If you want to argue about the performance you'll have to bring profiling results, not just theory to the table here.
> 
> I am just discussing the design, I did not even look at the code.
> 
> I can do a benchmark to show that `PFB_RB_FIND(&ctx->fb_tree, &lookup);` is slower than `ctx->fb_array[gv->idx];` but is this what we want to achieve ? (optimize speed of `PrefilterFlowbitMatch` )

I mean of this implementation with a real world use case.

**Comment by catenacyber** [comment]
For the "important corner case", am I correct to say that a dumb implementation of prefilter without this corner case will make 

```
alert http any any -> any any (http.uri; content:"test"; flowbits:set,uritest; sid:11;)
alert http any any -> any any (http.method; content:"HEAD"; flowbits:isset,uritest; prefilter; sid:12;)
```

with a packet that is `HEAD /test HTTP/1.1\n\n`

not trigger rule 12 when it should (and does trigger without the prefilter keyword), because rule 12 will be skipped at prefilter stage ?

Does it make sense for rule writers to prefilter on a flowbit that will be set during rules evaluation ?
But I guess we cannot know, based on the rules only, when one flowbit will be set (same packet, or strictly before its `isset` rule evaluation)

File: src/detect.c [hunk 8]
```diff
@@ -861,6 +871,51 @@ static inline void DetectRulePacketRules(
             }
         }
         AlertQueueAppend(det_ctx, s, p, txid, alert_flags);
+
+        if (det_ctx->post_rule_work_queue.len > 0) {
```
Comment (catenacyber) [inline src/detect.c:875]: Should it be a `while` ?

Like for
```
alert http any any -> any any (http.uri; content:"test"; flowbits:set,uritest; sid:11;)
alert http any any -> any any (http.method; content:"HEAD"; flowbits:isset,uritest; prefilter; flowbits:set,headtest; sid:12;)
alert http any any -> any any (http.ua; content:"toto"; flowbits:isset, headtest; prefilter; sid:13;)
```

1. Prefilter removes sids 12 and 13 because flowbits are not set
2. Rule evaluation makes sid 11 match and set flowbit uritest
3. We run post-"prefilter" for flowbit uritest, get sid 12
4. Rule evaluation makes sid 12 match and set flowbit headtest
5. We run post-"prefilter" again but for flowbit headtest, get sid 13
6. Rule evaluation makes sid 13 match and we have no more new flowbits set

File: src/detect-engine-prefilter.c (hunk 1 above)
Comment (victorjulien) [reply src/detect-engine-prefilter.c:393]: Currently just one I think.

File: src/detect-engine-prefilter.h (hunk 2 above)
Comment (victorjulien) [reply src/detect-engine-prefilter.h:1]: see how long this branch as been going? ;)

**Comment by victorjulien** [comment]
> I wonder how much of this need is covered by #12646 ...

Partly, but flowbits have a much wider scope. Ppl build state machines with them.

File: src/detect.c (hunk 8 above)
Comment (victorjulien) [reply src/detect.c:875]: I don't see why it should be a while. It's run a single time per signature that has this functionality enabled.

**Comment by victorjulien** [comment]
> For the "important corner case", am I correct to say that a dumb implementation of prefilter without this corner case will make
> 
> ```
> alert http any any -> any any (http.uri; content:"test"; flowbits:set,uritest; sid:11;)
> alert http any any -> any any (http.method; content:"HEAD"; flowbits:isset,uritest; prefilter; sid:12;)
> ```
> 
> with a packet that is `HEAD /test HTTP/1.1\n\n`
> 
> not trigger rule 12 when it should (and does trigger without the prefilter keyword), because rule 12 will be skipped at prefilter stage ?
> 
> Does it make sense for rule writers to prefilter on a flowbit that will be set during rules evaluation ? But I guess we cannot know, based on the rules only, when one flowbit will be set (same packet, or strictly before its `isset` rule evaluation)

It should not be affected by prefilter or not. Sid 11 will be evaluated first, because rule sorting puts it before 12 due to the flowbit set.

In non-prefilter, sid 12 would be evaluated, and in the evaluation we check if the flowbit "uritest" was set.

In the prefilter case sid 11 "knows" which sids depend on it, and it will add those dependencies for evaluation to the list of sig candidates. So without the match of sid 11, sid 12 would not have been evaluated. But now it is.

File: src/flow-bit.c [hunk 9]
```diff
@@ -91,13 +91,18 @@ void FlowBitUnset(Flow *f, uint32_t idx)
     FlowBitRemove(f, idx);
 }
 
-void FlowBitToggle(Flow *f, uint32_t idx)
+/**
+ * \retval bool true if bit is set, false is unset
+ */
+bool FlowBitToggle(Flow *f, uint32_t idx)
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/flow-bit.c:107]: yes

**Comment by catenacyber** [comment]
> > I wonder how much of this need is covered by #12646 ...
> 
> Partly, but flowbits have a much wider scope. Ppl build state machines with them.

Sure, but I can feel people doing flowbits because they do not have bidirectional rules, and then complain that the second rule does not have a good fast_pattern...

**Comment by catenacyber** [comment]
To be closed in favor of https://github.com/OISF/suricata/pull/12681 ?

File: src/detect.c (hunk 8 above)
Comment (catenacyber) [reply src/detect.c:875]: Indeed because the loop is done one level up as `for (uint32_t i = 0; i < array_idx; i++) {` and we just keep increasing the array

**Comment by catenacyber** [comment]
> In the prefilter case sid 11 "knows" which sids depend on it, and it will add those dependencies for evaluation to the list of sig candidates. So without the match of sid 11, sid 12 would not have been evaluated. But now it is.

You mean that is what your smart PR does with `PrefilterFlowbitPostRuleMatch`, not a dumb implementation of prefilter without the corner case, right ?

**Comment by catenacyber** [comment]
Should be closed in favor of https://github.com/OISF/suricata/pull/12681 ?
